url8c14fd53ee_luca89_itaco_tgz-pJ8-main_RunnerJ8/src/compilatore/JasminTarget.java

```diff
@@ -0,0 +1,636 @@
+		if (tabellaSimboliCorrente.containsKey(nome)) {
+			return tabellaSimboliCorrente.get(nome);
+		} else
+			throw new EccezioneSemantica(String.format(
+					"La variabile %s non è stata definita", nome));
+	}
+
+	protected int numeroVariabili() throws EccezioneSemantica {
+		return tabellaSimboliCorrente.size();
+	}
+
+	protected void pushScope() throws EccezioneSemantica {
+		tabellaSimboliCorrente = new HashMap<String, Integer>();
+		contatoreVariabiliGlobali = contatoreVariabiliCorrente;
+		contatoreVariabiliCorrente = 0;
+	}
+
+	protected void popScope() throws EccezioneSemantica {
+		tabellaSimboliCorrente = tabellaSimboliGlobale;
+		contatoreVariabiliCorrente = contatoreVariabiliGlobali;
+	}
+
+	private PrintStream outputFile;
+	private String className;
+	private ByteArrayOutputStream buffer;
+	private PrintStream bufferOutput;
+
+	/**
+	 * Costruttore che accetta il nome della classe e lo stream di output in cui
+	 * si dovrà salvare il codice target
+	 * 
+	 * @param className
+	 * @param output
+	 * @throws IOException
+	 */
+	public JasminTarget(String className, OutputStream output)
+			throws IOException {
+		this.outputFile = new PrintStream(output);
+		this.className = className;
+		initBuffer();
+		scriviContenutoStub("initStub.j", "%className", className);
+	}
+
+	/*
+	 * Utility per inizializzare e svuotare un buffer. Tutto il codice viene
+	 * prima scritto in questo buffer per consentire il calcolo del numero di
+	 * variabili da utilizzare nel programma e quindi settare correttamente i
+	 * valori di .limit e .stack
+	 */
+	private void initBuffer() {
+		buffer = new ByteArrayOutputStream();
+		bufferOutput = new PrintStream(buffer);
+	}
+
+	private void svuotaBuffer() {
+		byte[] ret = buffer.toByteArray();
+		outputFile.write(ret, 0, ret.length);
+		initBuffer();
+	}
+
+	/*
+	 * Scrive nel linguaggio target delle parti di codice indipendenti dallo
+	 * specifico programma
+	 */
+	private void scriviContenutoStub(String stubFileName,
+			String replacePattern, String replaceString) throws IOException {
+		/*
+		 * Copio il contenuto dello stub nel file di output
+		 */
+		InputStreamReader stub = new InputStreamReader(JasminTarget.class.getResourceAsStream("/compilatore/stub/"+stubFileName));
+		StringBuilder sb = new StringBuilder();
+		int charsRead = 0;
+		char[] fileData = new char[1024];
+
+		while (stub.ready()) {
+			charsRead = stub.read(fileData);
+			sb.append(fileData, 0, charsRead);
+		}
+		String stubReplaced = sb.toString().replaceAll(replacePattern,
+				replaceString);
+		outputFile.print(stubReplaced);
+	}
+
+	/*
+	 * Scrive nel linguaggio target delle parti di codice indipendenti dallo
+	 * specifico programma
+	 */
+	@SuppressWarnings("unused")
+	private void scriviContenutoStub(String stubFileName) throws IOException {
+		/*
+		 * Copio il contenuto dello stub nel file di output
+		 */
+		InputStream stub = JasminTarget.class.getResourceAsStream("/compilatore/stub/"+stubFileName);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the following resource: *stub*. Execution paths that do not contain closure statements were detected. Ensure *stub* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).